### PR TITLE
Adress issues by bzbarsky

### LINKED ISSUE #72: replace(text) should throw more informative exception when @import fails to load
The definition of [`replace(text)`](https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-replace) currently says that the promise is rejected with `NotAllowedError` if an `@import`-ed stylesheet fails to load.  This is the same error used (as a thrown exception, rather than promise rejection) in a number of other cases.  It seems like a more informative exception should be used for load failures -- perhaps `NetworkError` or maybe even more than one (e.g., `NetworkError`, `SecurityError`) depending on what happened.

### LINKED ISSUE #73: `[Constructor]` can't be used on partial interfaces
`[Constructor]` is defined in Web IDL to only be allowed on interfaces, not partial interfaces.  It looks like at least some browsers' IDL parsers don't enforce this, unfortunately, and make up some sort of semantics for what it means on a partial interface.

Either the `[Constructor]` needs to be moved into CSSOM or Web IDL needs to be changed to allow `[Constructor]` on partial interfaces and specify how it works.

### LINKED ISSUE #74: Wording around how to find a sheet's associated document is confusing
The spec says:

> with location set to the base URL of the Document for the Window where this constructor is called on

What is "where this constructor is called on"?   Especially when a constructor is moved from one window to another (which is totally a thing script can do)?  I suspect what you want is:

> with location set to the base URL of the associated Document of the current global.

where "associated" links to https://html.spec.whatwg.org/multipage/window-object.html#concept-document-window and "current global" links to https://html.spec.whatwg.org/multipage/webappapis.html#current-global-object

### LINKED ISSUE #75: Might be good to more clearly explain that the input MediaList gets cloned
Right now the spec says:

> assign a copy of the value of the attribute as sheet’s media

I assume this means the given MediaList gets cloned, but I can imagine someone being confused about it.  Presumably we would have test coverage that makes sure people don't screw this up.

Alternately, we could explicitly say that if _options_' "media" attribute is a `MediaList` then we should https://drafts.csswg.org/cssom-1/#serialize-a-media-query-list and then https://drafts.csswg.org/cssom-1/#create-a-medialist-object from the resulting string.  I believe that should be observably identical to the desired behavior, right?

### LINKED ISSUE #76: alternate flag should be specified when creating the sheet (editorial)
I _think_ CSSOM conceptually treats the "alternate flag" as immutable.  Hence the "specified when created" thing.  All invocations of https://drafts.csswg.org/cssom-1/#create-a-css-style-sheet in CSSOM explicitly say whether the flag should be set...

It might be best to write out the creation invocation in this spec as a list, more like the ones in CSSOM, which would make it clearer what all the things are being set to.

### LINKED ISSUE #77: Description for "constructed flag" could be a bit clearer
For the last sentence of the description, I would just say:

> Signifies whether this stylesheet was created by invoking the IDL-defined constructor.

and avoid talking about two different functions when there is only one function involved, etc.

### LINKED ISSUE #78: Typo in description of "disallow modification flag"
It says "modification ... are not allowed".  It should either be "is not allowed" or "modifications".

While we are here, it should be "modification of", not "modification to"...

### LINKED ISSUE #79: "constructor document" description is not very grammatical
Specifically, "is originally constructed on".

Probably better:

> The Document a constructed sheet is associated with.  Null by default.  Only non-null for sheets that have  "constructed flag" set.

and leave it at that.

### LINKED ISSUE #80: Might be better to not include function signatures when talking about functions
In https://wicg.github.io/construct-stylesheets/#modifying-constructed-stylesheets we have:

> like insertRule(rule[, index]) or deleteRule(index), or replace(text)

Including the arguments makes this harder to read, imo, and can get out of date if the linked-to method definitions change.  I'd just use the names here without signatures.

### LINKED ISSUE #82: Modifying Constructed Stylesheets should have a note that it's explicitly monkeypatching CSSOM
Probably with a link to the "merge into CSSOM" issue.

### LINKED ISSUE #85: Various grammar/phrasing issues in Modifying Constructed Stylesheets
In https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-insertrule step 1: "Let sheet be the context object" linking to https://dom.spec.whatwg.org/#context-object

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-insertrule step 1: missing "a" after "throw".

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-insertrule step 2: two flags should be "are", not "is".  Is there a reason to check for the "constructed flag" here?  Seems like if "disallowe modification" is set we should just throw.

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-insertrule step 3: Missing "a" after "throw".

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-deleterule step 1: Again, use "context object".

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-deleterule step 2: Missing "a" after "throw".

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-deleterule step 2: Why are we checking both flags?  Why not just the "disallow modification" flag?

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-replace step 1: "context object"

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-replace step 5 substep 2: If a load fails, why is a "NotAllowedError" exception what we reject with?  Seems a bit odd.  "NetworkError" would seem more appropriate.  And it should be "a NetworkError exception".

https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-replacesync step 1: "context object"

### LINKED ISSUE #86: replaceSync last steps seem a bit confused
https://wicg.github.io/construct-stylesheets/#dom-cssstylesheet-replacesync step 6 says to return sheet, but the method is declared as returning "void".

Step 4 modifies the sheet before step 5 would throw an exception.

This looks like a holdover from a factory method.  Presumably step 4 should parse a list of rules.  Then if a list of rules is returned step 5 would check it for `@import` rules and throw as needed.  Step 6 would assign the sheet's rule list.

### LINKED ISSUE #87: adoptedStyleSheets setter has some confusing verbiage
https://wicg.github.io/construct-stylesheets/#dom-documentorshadowroot-adoptedstylesheets step 2 says:

> If any entry of adopted has its constructed flag not set (e.g. it’s not made by factory methods to construct stylesheets)

Except there are no factory methods anymore.  Probably better to just say:

> If any entry of adopted does not have its "constructed flag" set

### PR DESCRIPTION
**Pull request by rakina**
Thanks @bzbarsky for the issues filed :)

Fixes #72, fixes #73, fixes #74, fixes #75, fixes #76, fixes #77, fixes #78, fixes #79, fixes #80, fixes #82, fixes #82, fixes #85, fixes #86, fixes #87.

Will merge soon if there's no complaints.

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/construct-stylesheets/pull/88.html" title="Last updated on Dec 21, 2018, 2:18 AM UTC (ea850d5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/construct-stylesheets/88/aeafe52...ea850d5.html" title="Last updated on Dec 21, 2018, 2:18 AM UTC (ea850d5)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -72,74 +73,83 @@ dictionary CSSStyleSheetInit {
...
+<p class="note">
+Note that we're explicitly monkeypatching CSSOM. We are working on [merging this spec into CSSOM](https://github.com/w3c/csswg-drafts/issues/3433) soon.
+</p>
 
 <dl>
     <dt><dfn method for=CSSStyleSheet>insertRule(|rule|, |index|)</dfn></dt>
     <dd>
-    	1. Let |sheet| be the stylesheet on which this function is called on.
-		2. If |sheet|'s [=constructed flag=] and [=disallow modification flag=] is set, throw "{{NotAllowedError}}" {{DOMException}}.
+    	1. Let |sheet| be the [=context object=].
```
Comment (domenic) [inline index.bs]: We're trying to move away from "context" object. Instead say `this {{CSSStyleSheet}} object`.

File: index.bs [hunk 2]
```diff
@@ -148,7 +158,7 @@ After construction, constructed stylesheets can be modified using rule modificat
 			2. Wait for loading of <a>@import</a> rules in |rules| and any nested <a>@import</a>s from those rules (and so on).
 				* If any of them failed to load, [=terminate=] fetching of the remaining <a>@import</a> rules,  and [=queue a task=] on the [=networking task source=] to perform the following steps:
                     1. Unset |sheet|'s [=disallow modification flag=].
-                    2. Reject |promise| with reason set to "{{NotAllowedError}}" {{DOMException}}.
+                    2. Reject |promise| with reason set to "{{NetworkError}}" {{DOMException}}.
```
Comment (domenic) [inline index.bs]:
```suggestion
                     2. Reject |promise| with a "{{NetworkError}}" {{DOMException}}.
```

**Review by domenic — APPROVED** [review]
LGTM with a couple nits. You may also want to rebase.

File: index.bs [hunk 3]
```diff
@@ -72,74 +73,83 @@ dictionary CSSStyleSheetInit {
...
+		Only non-null for stylesheets that have [=constructed flag=] set.
 	</dd>
 </dl>
 
 Modifying Constructed Stylesheets {#modifying-constructed-stylesheets}
 =============================
 
-After construction, constructed stylesheets can be modified using rule modification methods like <a spec=cssom-1 for="CSSStyleSheet">insertRule(rule[, index])</a> or <a spec=cssom-1 for="CSSStyleSheet">deleteRule(index)</a>, or {{replace(text)}} and {{replaceSync(text)}} if the sheet's [=disallow modification flag=] is not set. If those methods are called when the sheet's [=disallow modification flag=] is set, or <a spec=cssom-1 for="CSSStyleSheet">insertRule(rule)</a> is used to add an <a>@import</a> rule, a "{{NotAllowedError}}" {{DOMException}} will be thrown as detailed in the below algorithms. 
+After construction, constructed stylesheets can be modified using rule modification methods like [[cssom/#dom-cssstylesheet-insertrule|insertRule]] or [[cssom/#dom-cssstylesheet-deleterule|deleteRule]], or {{replace(text)}} and {{replaceSync(text)}}. The behavior of those methods are affected by the [=disallow modification flag=] and [=constructed flag=] as described below.
```
Comment (tabatkins) [inline index.bs]: This isn't the linking syntax you want. `{{CSSStyleSheet/insertRule()}}` will correctly link to the method. (You only need to supply arglists if the method has multiple overrides; if not, you can always autolink it with just `()`.)

File: index.bs [hunk 4]
```diff
@@ -72,74 +73,83 @@ dictionary CSSStyleSheetInit {
...
-			with <a spec=cssom>location</a> set to the <a spec=html>base URL</a> of the {{Document}} for the {{Window}} where this constructor is called on,
-			no <a spec=cssom>parent CSS style sheet</a>,
-			no <a spec=cssom>owner node</a>,
-			no <a spec=cssom>owner CSS rule</a>,
-			and <a spec=cssom>title</a> set to the {{CSSStyleSheetInit/title}} attribute of |options|.
-		2.	Set |sheet|'s origin-clean flag.
-			Set |sheet|'s [=constructed flag=].
-			Set |sheet|'s [=constructor document=] to the {{Document}} for the {{Window}} where this constructor is called on.
-		3. If the {{CSSStyleSheetInit/media}} attribute of |options| is a string,
+		1. Construct a new <a for="/" interface>CSSStyleSheet</a> object |sheet| with the following properties:
```
Comment (tabatkins) [inline index.bs]: Why are you using an explicit `<a>` here? `{{CSSStyleSheet}}` works fine and is more idiomatic.

File: index.bs (hunk 3 above)
Comment (rakina) [reply index.bs]: Oh, thanks!

File: index.bs (hunk 4 above)
Comment (rakina) [reply index.bs]: It's complaining for ambiguity if I do that:
Local references:
spec:construct-stylesheets-1; type:constructor; for:CSSStyleSheet; text:CSSStyleSheet()
for-less references:
spec:cssom-1; type:interface; for:/; text:CSSStyleSheet
spec:cssom-1; type:interface; for:/; text:CSSStyleSheet

Comment (tabatkins) [reply index.bs]: Ah, indeed! You can specify that the `for` is `/` with just `{{/CSSStyleSheet}}`, then. ^_^

Comment (rakina) [reply index.bs]: oh, neat!

**Pull request merged**